pkg/costattribution/active_tracker.go

```diff
@@ -0,0 +1,188 @@
+}
+
+func newActiveSeriesTracker(userID string, trackedLabels []string, limit int, cooldownDuration time.Duration, logger log.Logger) *ActiveSeriesTracker {
+	// Create a map for overflow labels to export when overflow happens
+	overflowLabels := make([]string, len(trackedLabels)+2)
+	for i := range trackedLabels {
+		overflowLabels[i] = overflowValue
+	}
+
+	overflowLabels[len(trackedLabels)] = userID
+	overflowLabels[len(trackedLabels)+1] = overflowValue
+
+	ast := &ActiveSeriesTracker{
+		userID:           userID,
+		labels:           trackedLabels,
+		maxCardinality:   limit,
+		observed:         make(map[string]*atomic.Int64),
+		logger:           logger,
+		overflowLabels:   overflowLabels,
+		cooldownDuration: cooldownDuration,
+	}
+
+	variableLabels := slices.Clone(trackedLabels)
+	variableLabels = append(variableLabels, tenantLabel, "reason")
+
+	ast.activeSeriesPerUserAttribution = prometheus.NewDesc("cortex_ingester_attributed_active_series",
+		"The total number of active series per user and attribution.", variableLabels[:len(variableLabels)-1],
+		prometheus.Labels{trackerLabel: defaultTrackerName})
+
+	return ast
+}
+
+func (at *ActiveSeriesTracker) hasSameLabels(labels []string) bool {
+	return slices.Equal(at.labels, labels)
+}
+
+func (at *ActiveSeriesTracker) Increment(lbls labels.Labels, now time.Time) {
+	if at == nil {
+		return
+	}
+	buf := bufferPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer bufferPool.Put(buf)
+	at.fillKeyFromLabels(lbls, buf)
+
+	at.observedMtx.RLock()
+	as, ok := at.observed[string(buf.Bytes())]
+	if ok {
+		as.Inc()
+		at.observedMtx.RUnlock()
+		return
+	}
+
+	if !at.overflowSince.IsZero() {
+		at.overflowCounter.Inc()
+		return
+	}
+	at.observedMtx.RUnlock()
+
+	at.observedMtx.Lock()
+	defer at.observedMtx.Unlock()
+	as, ok = at.observed[string(buf.Bytes())]
+	if ok {
+		as.Inc()
+		return
+	}
+
+	if !at.overflowSince.IsZero() {
+		at.overflowCounter.Inc()
+		return
+	}
+
+	if len(at.observed) >= at.maxCardinality {
+		at.overflowSince = now
+		at.overflowCounter.Inc()
+		return
+	}
+
+	at.observed[string(buf.Bytes())] = atomic.NewInt64(1)
+
+}
+
+func (at *ActiveSeriesTracker) Decrement(lbls labels.Labels) {
+	if at == nil {
+		return
+	}
+	buf := bufferPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer bufferPool.Put(buf)
+	at.fillKeyFromLabels(lbls, buf)
+	at.observedMtx.RLock()
+	as, ok := at.observed[string(buf.Bytes())]
+	if ok {
+		nv := as.Dec()
+		if nv > 0 {
+			at.observedMtx.RUnlock()
+			return
+		}
+		at.observedMtx.RUnlock()
+		at.observedMtx.Lock()
+		as, ok := at.observed[string(buf.Bytes())]
+		if ok && as.Load() == 0 {
+			// use buf.String() instead of string(buf.Bytes()) to fix the lint issue
+			delete(at.observed, buf.String())
+		}
+		at.observedMtx.Unlock()
+		return
+	}
+	at.observedMtx.RUnlock()
+
+	at.observedMtx.RLock()
+	if !at.overflowSince.IsZero() {
+		at.overflowCounter.Dec()
+		return
+	}
+	defer at.observedMtx.RUnlock()
```

Comment:
This should be before the `if`. We're not unlocking in the previous return statement.

Reply:
addressed here 80b64dc69133148805220b4d7638daaf95389676